## Heroes Of Pymoli


In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [2]:
pri_dic = {'Total Players':purchase_data['SN'].nunique()}
pd.DataFrame(pri_dic,index= [0])

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
pri_dic = {
    'Number of Unique Items':purchase_data['Item ID'].nunique(),
    'Average Price':'$'+str(round(purchase_data['Price'].mean(),2)),
    'Number of Purchases':len(purchase_data['Purchase ID']),
    'Total Revenue':'$'+str(purchase_data['Price'].sum())
}
pd.DataFrame(pri_dic,index= [0])

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780      $2379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
purchase_data_new = purchase_data[['SN','Gender']].drop_duplicates()
temp = purchase_data_new.groupby(['Gender'])['SN'].count().to_frame()
temp.reset_index(inplace=True)
temp_col = ['gender','Total Count']
temp.columns=temp_col
temp['Percentage of Players'] = temp['Total Count']/len(purchase_data_new['Gender'])
temp['Percentage of Players'] = temp['Percentage of Players'].apply(lambda x: '%.2f%%' % (x*100))
temp.set_index(['gender'],inplace=True)

In [5]:
temp

Total Count Percentage of Players
gender                                                  
Female                          81                14.06%
Male                           484                84.03%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [6]:
temp_1 = purchase_data.groupby(['Gender'])['Purchase ID'].count().to_frame()
temp_1.reset_index(inplace=True)
temp_1_col = ['Gender','Purchase Count']
temp_1.columns=temp_1_col

temp_2 = purchase_data.groupby(['Gender'])['Price'].mean().to_frame()
temp_2.reset_index(inplace=True)
temp_2_col = ['Gender','Average Purchase Price']
temp_2.columns=temp_2_col
temp_2['Average Purchase Price'] = temp_2['Average Purchase Price'].apply(lambda x: '$%.2f' % (x))

temp_3 = purchase_data.groupby(['Gender'])['Price'].sum().to_frame()
temp_3.reset_index(inplace=True)
temp_3_col = ['Gender','Total Purchase Value']
temp_3.columns=temp_3_col
temp_3['Total Purchase Value'] = temp_3['Total Purchase Value'].apply(lambda x: '$%.2f' % (x))

temp_4 = purchase_data.groupby(['Gender'])['Price'].sum().to_frame()
temp_4.reset_index(inplace=True)
temp_4_col = ['Gender','Avg Total Purchase per Person']
temp_4.columns=temp_4_col
for i in range(3):
    temp_4.iloc[i,1] = temp_4.iloc[i,1]/temp.iloc[i,0]
temp_4['Avg Total Purchase per Person'] = temp_4['Avg Total Purchase per Person'].apply(lambda x: '$%.2f' % (x))

temp_all = pd.merge(temp_1,temp_2)
temp_all = pd.merge(temp_all,temp_3)
temp_all = pd.merge(temp_all,temp_4)
temp_all.set_index(['Gender'],inplace=True)

In [7]:
temp_all

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [8]:

je_cut =[0,9,14,19,24,29,34,39,1000] 
je_label = ['<10','10-14','15-19','20-24','25-29','30-34','35-39','40+']  
purchase_data['Age Ranges'] = pd.cut(purchase_data['Age'],je_cut,labels= je_label)

In [9]:
purchase_data_new = purchase_data[['Age Ranges','SN','Age']].drop_duplicates()
temp_age = purchase_data_new.groupby(['Age Ranges'])['SN'].count().to_frame()
temp_age.reset_index(inplace=True)
temp_age_col = ['Age Ranges','Total Count']
temp_age.columns=temp_age_col
temp_age['Percentage of Players'] = temp_age['Total Count']/len(purchase_data_new)
temp_age['Percentage of Players'] = temp_age['Percentage of Players'].apply(lambda x: '%.2f%%' % (x*100))
temp_age.set_index(['Age Ranges'],inplace=True)

In [10]:
temp_age

Total Count Percentage of Players
Age Ranges                                   
<10                  17                 2.95%
10-14                22                 3.82%
15-19               107                18.58%
20-24               258                44.79%
25-29                77                13.37%
30-34                52                 9.03%
35-39                31                 5.38%
40+                  12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [11]:
temp_age_1 = purchase_data.groupby(['Age Ranges'])['Purchase ID'].count().to_frame()
temp_age_1.reset_index(inplace=True)
temp_age_1_col = ['Age Ranges','Purchase Count']
temp_age_1.columns=temp_age_1_col

temp_age_2 = purchase_data.groupby(['Age Ranges'])['Price'].mean().to_frame()
temp_age_2.reset_index(inplace=True)
temp_age_2_col = ['Age Ranges','Average Purchase Price']
temp_age_2.columns=temp_age_2_col
temp_age_2['Average Purchase Price'] = temp_age_2['Average Purchase Price'].apply(lambda x: '$%.2f' % (x))

temp_age_3 = purchase_data.groupby(['Age Ranges'])['Price'].sum().to_frame()
temp_age_3.reset_index(inplace=True)
temp_age_3_col = ['Age Ranges','Total Purchase Value']
temp_age_3.columns=temp_age_3_col
temp_age_3['Total Purchase Value'] = temp_age_3['Total Purchase Value'].apply(lambda x: '$%.2f' % (x))

temp_age_4 = purchase_data.groupby(['Age Ranges'])['Price'].sum().to_frame()
temp_age_4.reset_index(inplace=True)
temp_age_4_col = ['Age Ranges','Avg Total Purchase per Person']
temp_age_4.columns=temp_age_4_col
for i in range(len(temp_age_4)):
    temp_age_4.iloc[i,1] = temp_age_4.iloc[i,1]/temp_age.iloc[i,0]
temp_age_4['Avg Total Purchase per Person'] = temp_age_4['Avg Total Purchase per Person'].apply(lambda x: '$%.2f' % (x))

temp_age_all = pd.merge(temp_age_1,temp_age_2)
temp_age_all = pd.merge(temp_age_all,temp_age_3)
temp_age_all = pd.merge(temp_age_all,temp_age_4)
temp_age_all.set_index(['Age Ranges'],inplace=True)

In [12]:
temp_age_all

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05             $1114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [13]:
a = purchase_data.groupby(['SN']).agg({"Purchase ID":['count'],"Price":['mean','sum']})
temp_a = a['Purchase ID']
temp_a['Average Purchase Price'] = a['Price']['mean']
temp_a['Total Purchase Value'] = a['Price']['sum']
temp_a.columns = ['Purchase Count','Average Purchase Price','Total Purchase Value']
temp_a.sort_values(by='Total Purchase Value',ascending=False,inplace=True)
temp_a['Average Purchase Price'] = temp_a['Average Purchase Price'].apply(lambda x: '$%.2f' % (x))
temp_a['Total Purchase Value'] = temp_a['Total Purchase Value'].apply(lambda x: '$%.2f' % (x))

C:\Users\godlo\AppData\Local\Temp/ipykernel_5628/541852456.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp_a['Average Purchase Price'] = a['Price']['mean']


In [14]:
temp_a.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.41               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [15]:
a = purchase_data.groupby(['Item ID','Item Name']).agg({"Purchase ID":['count'],"Price":['mean','sum']})
temp_a = a['Purchase ID']
temp_a['Item Price'] = a['Price']['mean']
temp_a['Total Purchase Value'] = a['Price']['sum']
temp_a.columns = ['Purchase Count','Item Price','Total Purchase Value']
temp_a['Item Price'] = temp_a['Item Price'].apply(lambda x: '$%.2f' % (x))
temp_a['Total Purchase Value'] = temp_a['Total Purchase Value'].apply(lambda x: '$%.2f' % (x))
temp_a.sort_values(by='Purchase Count',ascending=False,inplace=True)

C:\Users\godlo\AppData\Local\Temp/ipykernel_5628/527577226.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp_a['Item Price'] = a['Price']['mean']


In [16]:
temp_a.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [17]:
a = purchase_data.groupby(['Item ID','Item Name']).agg({"Purchase ID":['count'],"Price":['mean','sum']})
temp_a = a['Purchase ID']
temp_a['Item Price'] = a['Price']['mean']
temp_a['Total Purchase Value'] = a['Price']['sum']
temp_a.columns = ['Purchase Count','Item Price','Total Purchase Value']
temp_a.sort_values(by='Total Purchase Value',ascending=False,inplace=True)
temp_a['Item Price'] = temp_a['Item Price'].apply(lambda x: '$%.2f' % (x))
temp_a['Total Purchase Value'] = temp_a['Total Purchase Value'].apply(lambda x: '$%.2f' % (x))

C:\Users\godlo\AppData\Local\Temp/ipykernel_5628/1611052705.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp_a['Item Price'] = a['Price']['mean']


In [18]:
temp_a.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Trends Description

The largest group of gender players are male players counts for 484 in total , taking part of 84.03% which is more than half of the total numbers. Despite, the average purchase value stays similar despite Male group contains larger number of people which gives the largest number of total purchase amount. 

The age between 20-24 is the largest group of player almost counts for 44.79% in total and second is 15-19 which is only 18.58% of total players. The average purchase in age not differentiate too much between different groups, which 35-39 group is highest spending group, but age 20-24 gives largest total value due to the largest number of players.

The most popular items are Item ID 92, the Final Critic which gives total count for 13 purhcases and total value 59.99$. This number also gives it as the most profitable item meantime.